src/net_processing.cpp

```diff
@@ -1880,6 +1930,240 @@ void static ProcessOrphanTx(CConnman* connman, std::set<uint256>& orphan_work_se
+ * case, block for a short time until the filter index is updated, then retry the lookup.
+ */
+static bool QueryFilterIndexWithRetry(BaseIndex* index, bool& in_sync, std::function<bool()> fn)
+{
+    while (!fn()) {
+        if (in_sync) {
+            return false;
+        }
+        if (!index->BlockUntilSyncedToCurrentChain()) {
+            return error("%s is not ready yet", index->GetName());
+        }
+        in_sync = true;
+    }
+    return true;
+}
+
+static bool ProcessGetCFilters(CNode* pfrom, CDataStream& vRecv, const CChainParams& chain_params,
+                               CConnman* connman)
+{
+    uint8_t filter_type_ser;
+    uint32_t start_height;
+    uint256 stop_hash;
+
+    vRecv >> filter_type_ser >> start_height >> stop_hash;
+
+    BlockFilterType filter_type = static_cast<BlockFilterType>(filter_type_ser);
+
+    const CBlockIndex* stop_index;
+    BlockFilterIndex* filter_index;
+    if (!PrepareBlockFilterRequest(pfrom, chain_params, filter_type, start_height, stop_hash,
+                                   MAX_GETCFILTERS_SIZE, stop_index, filter_index)) {
+        // Return true because the issue with the invalid request has already been logged.
+        return true;
+    }
+
+    bool index_in_sync = false;
+
+    std::vector<BlockFilter> filters;
+    bool lookup_success = QueryFilterIndexWithRetry(
+        filter_index, index_in_sync,
+        [=, &filters]{ return filter_index->LookupFilterRange(start_height, stop_index, filters); }
+    );
+    if (!lookup_success) {
+        return error("Failed to find block filter in index: filter_type=%s, start_height=%d, stop_hash=%s",
+                     BlockFilterTypeName(filter_type), start_height, stop_hash.ToString());
+    }
+
+    for (const auto& filter : filters) {
+        CSerializedNetMsg msg = CNetMsgMaker(pfrom->GetSendVersion())
+            .Make(NetMsgType::CFILTER, filter);
+        connman->PushMessage(pfrom, std::move(msg));
+    }
+
+    return true;
+}
+
+static bool ProcessGetCFHeaders(CNode* pfrom, CDataStream& vRecv, const CChainParams& chain_params,
+                                CConnman* connman)
+{
+    uint8_t filter_type_ser;
+    uint32_t start_height;
+    uint256 stop_hash;
+
+    vRecv >> filter_type_ser >> start_height >> stop_hash;
+
+    BlockFilterType filter_type = static_cast<BlockFilterType>(filter_type_ser);
+
+    const CBlockIndex* stop_index;
+    BlockFilterIndex* filter_index;
+    if (!PrepareBlockFilterRequest(pfrom, chain_params, filter_type, start_height, stop_hash,
+                                   MAX_GETCFHEADERS_SIZE, stop_index, filter_index)) {
+        // Return true because the issue with the invalid request has already been logged.
+        return true;
+    }
+
+    bool index_in_sync = false;
+
+    uint256 prev_header;
+    if (start_height > 0) {
+        const CBlockIndex* prev_block = stop_index->GetAncestor(start_height - 1);
```

Comment:
nit: could be const CBlockIndex* **const** prev_block

Also, IDE notes about `start_height - 1` being implicitly converted to `int` (`GetAncestor()` takes int), maybe worth "documenting"/"warning" by using `static_cast`? Though maybe overkill, idk.